### PR TITLE
Allow multiple PV directory

### LINKED ISSUE #26: [storage-hostpath] Support multiple storage classes
It seems https://github.com/canonical/hostpath-provisioner doesn't have an issue tracker, so I'm posting here.

It would be nice to make the `hostpath-provisioner` of the `storage-hostpath` plugin respect multiple storage classes.

Example use-case: two volumes bind-mounted into the snap, one HDD and one SSD. Currently using the provisioner for one volume and manually mounting for the other, but it's a bit more tedious than it could be.

### PR DESCRIPTION
**Pull request by balchua**
Fixes https://github.com/canonical/microk8s-core-addons/issues/26

Allows one to specify different `StorageClass` to point to different host paths.

example:

``` yaml
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: slow-disk-hostpath
  annotations:
    storageclass.kubernetes.io/is-default-class: "true"
provisioner: microk8s.io/hostpath
parameters:
  pvDir: /slow-disk/ssd-storage
volumeBindingMode: WaitForFirstConsumer
reclaimPolicy: Retain
---
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: ssd-hostpath
provisioner: microk8s.io/hostpath
reclaimPolicy: Delete
parameters:
  pvDir: /fast-disk/ssd-storage
volumeBindingMode: WaitForFirstConsumer

### CODE REVIEW

File: hostpath-provisioner.go [hunk 1]
```diff
@@ -193,7 +193,13 @@ waitLoop:
 
 // Provision creates a storage asset and returns a PV object representing it.
 func (p *hostPathProvisioner) Provision(ctx context.Context, options controller.ProvisionOptions) (*v1.PersistentVolume, controller.ProvisioningState, error) {
-	path := path.Join(p.pvDir, fmt.Sprintf("%s-%s-%s", options.PVC.Namespace, options.PVC.Name, options.PVName))
+	pvDir := p.pvDir
+	storageClassPvDir, ok := options.StorageClass.Parameters["pvDir"]
+
+	if ok {
+		pvDir = storageClassPvDir
+	}
```
Comment (neoaggelos) [inline hostpath-provisioner.go]:
```suggestion
	if storageClassPvDir, ok := options.StorageClass.Parameters["pvDir"]; ok {
		pvDir = storageClassPvDir
	}
```

File: hostpath-provisioner.go [hunk 2]
```diff
@@ -84,7 +84,7 @@ func NewHostPathProvisioner(clientset *kubernetes.Clientset) controller.Provisio
 		klog.Fatal("env variable NODE_NAME must be set so that this provisioner can identify itself")
 	}
 
-	pvDir := os.Getenv("PV_DIR")
+	pvDir := os.Getenv("DEFAULT_PV_DIR")
```
Comment (neoaggelos) [inline hostpath-provisioner.go]: Let's not change the `PV_DIR` environment variable, backwards compatibility is much preferred here.

File: manifests/storage.yaml [hunk 3]
```diff
@@ -13,36 +13,25 @@ spec:
...
               valueFrom:
                 fieldRef:
                   fieldPath: spec.nodeName
-
-            - name: PV_DIR
+            - name: NAMESPACE
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.namespace
+            - name: DEFAULT_PV_DIR
```
Comment (neoaggelos) [inline manifests/storage.yaml]: i'd suggest reverting this; see above

```suggestion
            - name: DEFAULT_PV_DIR
```

File: manifests/storage.yaml [hunk 4]
```diff
@@ -13,36 +13,25 @@ spec:
...
   template:
     metadata:
       labels:
         k8s-app: hostpath-provisioner
-
     spec:
       containers:
         - name: hostpath-provisioner
-          image: kjackal/hostpath-provisioner:latest
+          image: balchu/hostpath-provisioner:1.1.1
```
Comment (neoaggelos) [inline manifests/storage.yaml]: let's revert this as well. even better, use the image from `cdkbot/hostpath-provisioner`

**Review by neoaggelos — CHANGES_REQUESTED** [review]
Nice work, left some comments.

Where do you think this should be documented?

File: hostpath-provisioner.go (hunk 2 above)
Comment (balchua) [reply hostpath-provisioner.go]: Ok 👍

File: manifests/storage.yaml (hunk 3 above)
Comment (balchua) [reply manifests/storage.yaml]: Ok 👍

**Comment by balchua** [comment]
@neoaggelos, 
2 places i can think of:

* MicroK8s website.
* Output of the enable script by providing samples.

**Comment by balchua** [comment]
Btw i do have a question on the hostpath location.  If the user mounts it outside `/var/snap/microk8s` and then uninstalled the snap.  It will be left out.  And will it work on strict confinement?

**Comment by neoaggelos** [comment]
> Btw i do have a question on the hostpath location. If the user mounts it outside `/var/snap/microk8s` and then uninstalled the snap. It will be left out.

Yes, this is correct, there is no way that MicroK8s can know about all these folders and clean them up.

> And will it work on strict confinement?

It should work out of the box for external storage, yes.

**Comment by neoaggelos** [comment]
Hi @balchua I'm quite interested to move this forward, especially with 1.25 around the corner. Have you had time to address the comments?

**Comment by balchua** [comment]
Let me try tonight. 😊

**Comment by balchua** [comment]
@neoaggelos 
suggestions resolved. i also tested it on `amd64` arch :) this one is good to go!
Thanks!

**Comment by balchua** [comment]
i will prolly update the documentation in MicroK8s docs site.

**Review by neoaggelos — APPROVED** [review]
LGTM, thanks @balchua!

**Comment by balchua** [comment]
Thanks @neoaggelos

**Pull request merged**